relayer/middleware/metrics.middleware.ts

```diff
@@ -1,45 +1,99 @@
-import { register } from "prom-client";
+import { Histogram, register } from "prom-client";
 import { Middleware } from "../compose.middleware";
-import { Context } from "../context";
 import { StorageContext } from "../storage/storage";
-import { Job } from "bullmq";
 import { Counter, Registry } from "prom-client";
 
-interface MetricsOpts {
+
+type MetricRecord = Record<string, string | number>;
+
+export interface MetricsOpts<C extends StorageContext> {
   registry?: Registry;
+  labels?: MetricLabelsOpts<C>;
+  processingTimeBuckets?: number[];
 }
 
-export function metrics(
-  opts: MetricsOpts = {},
-): Middleware<Context & { job?: Job }> {
+export class MetricLabelsOpts<C extends StorageContext> {
+  labelNames: string[];
+  customizer: (context: C) => Promise<MetricRecord>;
+
+  constructor(labelNames: string[] = [], customizer: (ctx: C) => Promise<MetricRecord>) {
+    this.labelNames = labelNames;
+    this.customizer = customizer;
+  }
+}
+
+export function metrics<C extends StorageContext>(
+  opts: MetricsOpts<C> = { },
+): Middleware<C> {
   opts.registry = opts.registry || register;
+
+
+  const getLabels = async (ctx: C, failed: boolean) => {
+    const defaultLabels = { status: failed ? "failed" : "succeeded" };
+
+    try {
+      const labels: MetricRecord = await (opts.labels?.customizer(ctx) ?? Promise.resolve({}));
```

Comment:
not acctionable comment:
I'm not 100% sure if `customizer` should support async functions :thinking: 
In general I'd try not to add async calls while building my metrics. Adding IO adds the possibility of a failure between the moment the event you are meassuring happens and the metrics are emitted, so you can't trust your metrics 100%
Making it async allows the user to make this decisition, which I think it's fine... and also the user gets the flexibility if they run into a case where they do need async processing before emitting metrics :)